subcommand/consul-init/command.go

```diff
@@ -0,0 +1,162 @@
+	"strings"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/hashicorp/consul-k8s/consul"
+	"github.com/hashicorp/consul-k8s/subcommand/common"
+	"github.com/hashicorp/consul-k8s/subcommand/flags"
+	"github.com/hashicorp/consul/api"
+	"github.com/mitchellh/cli"
+)
+
+const bearerTokenFile = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+const tokenSinkFile = "/consul/connect-inject/acl-token"
+
+type Command struct {
+	UI cli.Ui
+
+	flagACLAuthMethod   string // Auth Method to use for ACLs, if enabled
+	flagMeta            string // Flag for metadata to consul login.
+	flagBearerTokenFile string // Location of the bearer token.
+	flagTokenSinkFile   string // Location to write the output token.
+	numACLLoginRetries  int    // Number of times to attempt to retry ACL().Login(), default: 3.
+
+	flagSet *flag.FlagSet
+	http    *flags.HTTPFlags
+
+	consulClient *api.Client
+
+	sigCh chan os.Signal
+	once  sync.Once
+	help  string
+	cert  atomic.Value
+}
+
+func (c *Command) init() {
+	c.flagSet = flag.NewFlagSet("", flag.ContinueOnError)
+	c.flagSet.StringVar(&c.flagACLAuthMethod, "method", "",
+		"The name of the Kubernetes Auth Method to use for connectInjection if ACLs are enabled.")
+	c.flagSet.StringVar(&c.flagMeta, "meta", "",
+		"Metadata to be passed to consul login api.")
+	c.flagSet.StringVar(&c.flagBearerTokenFile, "bearer-token-file", bearerTokenFile,
+		"Token file to be used for consul login api.")
+	c.flagSet.StringVar(&c.flagTokenSinkFile, "token-sink-file", tokenSinkFile,
+		"Token sink file to be used for consul login api.")
+
+	c.http = &flags.HTTPFlags{}
+
+	flags.Merge(c.flagSet, c.http.Flags())
+	c.help = flags.Usage(help, c.flagSet)
+
+	// Default number of times to attempt ACL().Login()
+	c.numACLLoginRetries = 3
+
+	// Wait on an interrupt or terminate for exit, be sure to init it before running
+	// the controller so that we don't receive an interrupt before it's ready.
+	if c.sigCh == nil {
+		c.sigCh = make(chan os.Signal, 1)
+		signal.Notify(c.sigCh, syscall.SIGINT, syscall.SIGTERM)
+	}
+}
+
+func (c *Command) Run(args []string) int {
+	c.once.Do(c.init)
+	var err error
+	if err := c.flagSet.Parse(args); err != nil {
+		return 1
+	}
+
+	// Validate flags.
+	if c.flagACLAuthMethod == "" {
+		c.UI.Error("-method must be set")
+		return 1
+	}
+	if c.flagMeta == "" {
+		c.UI.Error("-meta must be set")
+		return 1
+	}
+	if c.flagBearerTokenFile == "" {
+		c.UI.Error("-bearer-token-file must be set")
+		return 1
+	}
+	if c.flagTokenSinkFile == "" {
+		c.UI.Error("-token-sink-file must be set")
+		return 1
+	}
+
+	// TODO: Add namespace support
+	if c.consulClient == nil {
+		cfg := api.DefaultConfig()
+		c.http.MergeOntoConfig(cfg)
+		c.consulClient, err = consul.NewClient(cfg)
+		if err != nil {
+			c.UI.Error("unable to get client connection")
+			return 1
+		}
+	}
+	ErrCh := make(chan error)
+	ExitCh := make(chan bool)
+	meta := map[string]string{"pod": strings.Split(c.flagMeta, "=")[1]}
+	go func() {
```

Comment:
This loop ignore any sigterms right now. I'm also not sure it needs to run in a go routine? Would something like this work:

```go
	retries := 0
	for {
		err = common.ConsulLogin(c.consulClient, c.flagBearerTokenFile, c.flagACLAuthMethod, c.flagTokenSinkFile, meta)
		if err == nil {
			break
		}
		retries++
		if retries == c.numACLLoginRetries {
			c.UI.Error("hit maximum retries for consul login")
			return 1
		}
		c.UI.Error(fmt.Sprintf("consul login failed; retrying: %s", err))

		select {
		case <-time.After(1 * time.Second):
			// retry loop
		case sig := <-c.sigCh:
			c.UI.Info(fmt.Sprintf("%s received, shutting down", sig))
			return 0
		}
	}
```

Reply:
Love it!

Reply:
What do you think about using `backoff.Retry` (example [here](https://github.com/hashicorp/consul-k8s/blob/125f7e37bc2ab16a7ed432849fbbfe13ff58e802/subcommand/get-consul-client-ca/command.go#L101-L115)). I think this package will already handle a lot the signals and make the logic much cleaner. You'd need to pass in `backoff.WithMaxRetries(backoff.NewConstantBackOff(1*time.Second), 3)` as the second argument so that it stops after 3 tries.

Reply:
@ishustava great idea, I like it and its much cleaner looking I'm not sure I understand how it handles signals. In `get-consul-client-ca` I don't see it either and we're not testing that command for any signal handling. Could you can show me? 🙇🏽

Reply:
What is the reason to have signal handling in this command?

In get-consul-client-ca, we don't handle interrupt signals because it doesn't start any goroutines, and so if an interrupt or a term signal is sent, it doesn't need to drain or stop any other processes that it's running. We also don't do it in other similar commands, e.g server-acl-init, probably for the same reason.

Reply:
We do loop for the login so wouldn't it be nice if you could delete the pod and it would exit quickly if it was in its `init`?

Reply:
I'm fine with not having it too if it's deemed too complicated.

Reply:
>We do loop for the login so wouldn't it be nice if you could delete the pod and it would exit quickly if it was in its init

If you delete a pod, then it should still exit quickly. If you have a program that doesn't handle interrupts and you send it an interrupt, it should still exit as soon as you send that signal because it's handled by golang and OS. In this command, we don't need any graceful shutdown behavior so I think it's ok to skip OS signal handling. Maybe I'm misunderstanding something?

Reply:
Ahh I didn't know that there was default signal handling that gets turned off by `signal.Notify` (https://golang.org/pkg/os/signal/).